Source/WTF/wtf/glib/RunLoopGLib.cpp

```diff
@@ -86,48 +85,73 @@ RunLoop::RunLoop()
 RunLoop::~RunLoop()
 {
     g_source_destroy(m_source.get());
+    m_shouldStop = true;
+}
+
+void RunLoop::runGLibMainLoopIteration(MayBlock mayBlock)
+{
+    auto* pollFunction = g_main_context_get_poll_func(m_mainContext.get());
+
+    static thread_local Vector<GPollFD, 32> pollFDs;
+    pollFDs.clear();
+
+    gint maxPriority = 0;
+    g_main_context_prepare(m_mainContext.get(), &maxPriority);
 
-    for (int i = m_mainLoops.size() - 1; i >= 0; --i) {
-        if (!g_main_loop_is_running(m_mainLoops[i].get()))
-            continue;
-        g_main_loop_quit(m_mainLoops[i].get());
+    pollFDs.grow(16);
+    auto* pollFDsData = pollFDs.mutableSpan().data();
+
+    gint timeoutInMilliseconds = 0;
+    auto numFDs = g_main_context_query(m_mainContext.get(), maxPriority, &timeoutInMilliseconds, pollFDsData, pollFDs.size());
+    if (static_cast<size_t>(numFDs) > pollFDs.size()) {
+        pollFDs.grow(numFDs);
+        numFDs = g_main_context_query(m_mainContext.get(), maxPriority, &timeoutInMilliseconds, pollFDsData, pollFDs.size());
```

Comment:
I still think we should iterate, another thread can attach a new source between the query calls

Reply:
https://github.com/WebKit/WebKit/pull/51678#discussion_r2405640323